### DISCUSSION TITLE
Unrelated but - This Project includes OTA with micro-python?  If so how can I implement it for my own project? Thanks

### DISCUSSION THREAD

**Discussion by tongclement** [arendst/Tasmota#12430] [Q&A] [answered]
Dear Tasmota Devs and the community,

Thanks for the great project. I have been redirected here by a comment in this video: https://www.youtube.com/watch?v=1pwqS_NUG7Q&lc=Ugx8zQJco2q2E-qhWeJ4AaABAg.9HEESdY255R9HRXEbKCufs 

as I was looking forward to implementing some form of OTA for my Esp32 Micropython project! 
Does anyone have any idea where I can reference for implementing some form of OTA for ESP32 running micro python? Thanks!

Clement

**Comment by barbudor** [comment, marked as answer]
Clement

Tasmota is a binary executable which implements an OTA process to update itself while running. On ESP32 (slighly different on ESP8266) the flash contains 2 executable blocks : one with the current active firmware, another one available for a new firmware. When a new firmware is sent to the device, Tasmota writes it to the free block, makes it the boot default and restart the ESP which then boot the new firmware.

In you case, you have to upload new files .py (or precompiled .pyc) to the ESP flash file system where MicroPython execute them. There was no built-in mecanism for that in MicroPython, you would have your application to include (like Tasmota does) a part of the web server to receive files and write them to the file system. Once new files have been updated, restart tge ESP and Micropython will execte new Python files. In a sense this is close to what Tasmota does to upload files to it's file system.
You may look in details the latest Micropython 1.16 which seems to include new features to ease upload of files.

> This release of MicroPython includes a new command-line tool called "mpremote", which is intended to be the main way to remotely control a MicroPython-based device from the command line. It features a serial terminal, filesystem access, support to mount a local directory on the remote device, and a macro language to define custom commands. This tool can be installed from PyPI via "pip3 install mpremote", and it works on Linux, Windows and Mac. As part of this, improvements were made to pyboard.py including opening serial ports in exclusive mode to more easily manage multiple devices.
